test/hotspot/jtreg/serviceability/jvmti/vthread/PopFrameTest/PopFrameTest.java

```diff
@@ -0,0 +1,252 @@
+
+    static int status = PASSED;
+    static boolean is_virtual = true;
+
+    static void setFailed(String msg) {
+        log("\nFAILED: " + msg);
+        status = FAILED;
+    }
+
+    static void throwFailed(String msg) {
+        log("\nFAILED: " + msg);
+        throw new RuntimeException("PopFrameTest failed!");
+    }
+
+    public static void main(String args[]) {
+        is_virtual = !(args.length > 0 && args[0].equals("platform"));
+        run();
+        if (status == FAILED) {
+            throwFailed("PopFrameTest!");
+        }
+        log("\nPopFrameTest passed");
+    }
+
+    public static void run() {
+        TestTask testTask = new TestTask();
+        Thread testTaskThread = null;
+        int errCode;
+
+        prepareAgent(TestTask.class);
+
+        log("\nMain #A: method A() must be blocked on entering a synchronized statement");
+        if (is_virtual) {
+            testTaskThread = Thread.ofVirtual().name("TestTaskThread").start(testTask);
+        } else {
+            testTaskThread = Thread.ofPlatform().name("TestTaskThread").start(testTask);
+        }
+        testTask.ensureStarted();
+
+        log("\nMain #A.1: unsuspended");
+        {
+            errCode = popFrame(testTaskThread);
+            if (errCode != THREAD_NOT_SUSPENDED) {
+                throwFailed("Main #A.1: expected THREAD_NOT_SUSPENDED instead of: " + errCode);
+            } else {
+                log("Main #A.1: got expected THREAD_NOT_SUSPENDED");
+            }
+
+            log("\nMain #A.2: suspended");
+            suspendThread(testTaskThread);
+            errCode = popFrame(testTaskThread);
+            if (errCode != JVMTI_ERROR_NONE) {
+                throwFailed("Main #A.2: expected JVMTI_ERROR_NONE instead of: " + errCode);
+            } else {
+                log("Main #A.2: got expected JVMTI_ERROR_NONE");
+            }
+            resumeThread(testTaskThread);
+            testTask.clearDoLoop();
+            testTask.sleep(5);
+        }
+
+        log("\nMain #B: method B() must be blocked in a breakpoint event handler");
+        {
+            ensureAtBreakpoint();
+
+            log("\nMain #B.1: unsuspended");
+            errCode = popFrame(testTaskThread);
+            if (errCode != THREAD_NOT_SUSPENDED) {
+                throwFailed("Main #B.1: expected THREAD_NOT_SUSPENDED instead of: " + errCode);
+            }
+            log("Main #B.1: got expected THREAD_NOT_SUSPENDED");
+
+            log("\nMain #B.2: suspended");
+            suspendThread(testTaskThread);
+            errCode = popFrame(testTaskThread);
+            if (errCode != JVMTI_ERROR_NONE) {
+                throwFailed("Main #B.2: expected JVMTI_ERROR_NONE");
+            }
+            log("Main #B.2: got expected JVMTI_ERROR_NONE");
+            resumeThread(testTaskThread);
+            notifyAtBreakpoint();
+
+            log("\nMain #B.3: unsuspended, call PopFrame on own thread");
+            ensureAtBreakpoint();
```

Comment:
Am I understand correctly, that test expect here to pop frame and immediately get to the same breakpoint?

Reply:
Yes, it is correct. The method `B()` is called twice in the method `run()` at lines `212` and `215`.